corehq/util/es/elasticsearch.py

```diff
@@ -2,63 +2,95 @@
 
 if settings.ELASTICSEARCH_MAJOR_VERSION == 1:
     import elasticsearch
-    from elasticsearch.exceptions import AuthorizationException
-    from elasticsearch import (
-        ConnectionError,
-        ConnectionTimeout,
-        Elasticsearch,
-        ElasticsearchException,
-        NotFoundError,
-        SerializationError,
-        ConflictError,
-        TransportError,
-        RequestError,
-    )
-    from elasticsearch.client import (
-        IndicesClient,
-        SnapshotClient,
-    )
-    from elasticsearch.helpers import bulk
 elif settings.ELASTICSEARCH_MAJOR_VERSION == 2:
     import elasticsearch2 as elasticsearch
-    from elasticsearch2.exceptions import AuthorizationException
-    from elasticsearch2 import (
-        ConnectionError,
-        ConflictError,
-        ConnectionTimeout,
-        Elasticsearch,
-        ElasticsearchException,
-        NotFoundError,
-        SerializationError,
-        TransportError,
-        RequestError,
-    )
-    from elasticsearch2.client import (
-        IndicesClient,
-        SnapshotClient,
-    )
-    from elasticsearch2.helpers import bulk
 elif settings.ELASTICSEARCH_MAJOR_VERSION == 7:
```

Comment:
Hey @dannyroberts I was trying to create a decorator to run ESs test on both ES versions by overriding a setting. Since these classes are loaded at the initial time, overriding the setting doesn't take effect. So, I tried to change the approach by implementing [this below](https://github.com/dimagi/commcare-hq/pull/27378/commits/702f5aff947d9ccd9612a1dd4e3acf20cbb71c4e#diff-ad8416e19381615f1d29199cd1726167R13). That works to some extent, but catching exceptions (such as ElasticsearchException) becomes an issue since this only works if the classes are instantiated but in a try-except check the check is done against the class itself. So this approach is also not working well.

Do you have any other ideas on how to make this work?

Reply:
I ended up creating [this decorator](https://github.com/dimagi/commcare-hq/pull/27378/commits/87d09ab71c32b7af83386b6f275eaebc699436a9#diff-156cccef95e5b965b2751ec341ade862R66) called `run_on_es2` to duplicate all ES tests to run on ES2 apart from the default ES7.

Reply:
@esoergel So, yeah, with this we will be able to test for both ES2 and ES7 in one build. (currently this PR is not yet setup to do that, but been doing it locally)

I had to use lot of magic to create this decorator to get around bunch of problems stated above. Any feedback is greatly welcome. Also @snopoke @millerdev 

(Note that this PR is still WIP and so please hold of on comments on other sectiosn apart from that decorator and the way it's designed to be used)

Reply:
The main issue with this approach is having to reload almost every module that loads the basic ES code blocks. But I don't have any other better approach so far.

Reply:
Considering a different approach. Please see https://dimagi.slack.com/archives/C435VLT2T/p1591363126131500